Payload Passthru Mode [WIP]

KSDaemon opened the pull request:
Describing and discussing payload transparency feature

ecorm commented on rfc/text/advanced/ap_50_payload_transparency.md:
```diff
@@ -0,0 +1,251 @@
+## Payload Transparency
```
"Transparent" means that the contents of the payload is visible. The terminology you need to use is "opaque payload" or "payload opacity".

https://www.merriam-webster.com/dictionary/transparent
https://www.merriam-webster.com/dictionary/opaque
https://www.merriam-webster.com/dictionary/opacity
https://english.stackexchange.com/questions/87177/difference-between-opacity-and-opaqueness

oberstet commented:
@ecorm sorry for having this wrong term creep on! we need a better name

@KSDaemon apparently this term is wrong / suggests the wrong meaning, so we need a different name and feature ID - Emile mentioned this previously and raised my attention to this, but of course all the existing experimental code in AB and CB are still on the wrong term. sorry, you couldn't know.

"have the router forward app payload without touching or caring": what's a good name for that?

KSDaemon commented:
haha! Feeling so stupid))))) Really))) Transparency.... And we are discussing encryption%)))

oberstet commented:
rgd having a new field to distinguish between different types of `enc_keyid`:

* the value must a be string, always and for any of the subtypes
* the value could be: a Hex-encoded string, a UUID encoded in standard format, an Ethereum address, a forward (or reverse) DNS name, ... 
* which type a (base)-string-typed value does indeed have can be

1. auto-detected (eg https://github.com/crossbario/autobahn-python/blob/9f425ff5792348c2f6dc3e7c3297c82f7fb5e793/autobahn/wamp/message.py#L231)
2. determined by `enc_algo`
3. determined by a new attribute `enc_kms` ("key management schema")
4. determined by a prefix, eg `enc_keyid = "xbr://0xe59C7418403CF1D973485B36660728a5f4A8fF9c"`

oberstet commented:
rather than "payload transparency", how about one of these?

1. "raw payload mode"
2. "unprocessed payload mode"
3. "zero-touch payload mode"
4. "zero-parse payload"

KSDaemon commented:
What came to my mind:
* Payload shadowing or Shadow Payload
* Payload opacity by @ecorm 
* Payload hiding

oberstet commented:
@ecorm 

rgd "Payload opacity", doesn't that suggest there would be a level or grade of opacity? which isn't the case. as when the mode is on, the router doesn't care at all, and if it is off, the router fully parses according to the wamp session serializer

how about "Opaque Payload Mode"?

oberstet commented:
rgd adding a be attribute `enc_kms` or a prefix to identify the key ID / management scheme to be used (and in use for the ID string format):

if we do that, then we should redefine `enc_algo` as well

because currently, `enc_algo` determines _both_ the actual cryptographic encryption algorithm (eg XSalsa20Poly1305 for XBR) _and_ the naming schema in use (UUIDs mapped to ETH addresses) ...

IOW: what is the advantage of using 2 attributes to identify those? I mean, the client needs to know/determine both anyways

counter argument: right, but then why do we use a separate attribute `enc_serializer` and not only have exactly _one_ to determine all of:

* crypto algo used
* key management scheme used
* serializer used (before encryption)

mmmh;)

ecorm commented:
> "have the router forward app payload without touching or caring": what's a good name for that?

Perhaps "opaque" isn't the best term because the router can still see the payload (even if it can't understand it when encrypted).

How about:
- Intact Payloads
- Pristine Payloads
- Protected Payloads
- Payload Conservation
- Immaculate Payloads

KSDaemon commented:
I like `Protected Payloads`.

oberstet commented:
awesome!! *Protected Payloads* or *Protected Payload Mode* works perfectly for me as well (not a native speaker anyways, so I lack the real clue about connotations .. anyways;) +1

KSDaemon commented:
Cool! I'll update docs to → `Protected Payload Mode`

oberstet commented:
btw, the list of all WAMP message types that can have `enc_algo` and `args[0] == payload|binary`:

```
(base) oberstet@intel-nuci7:~$ grep "Payload|binary]" scm/crossbario/autobahn-python/autobahn/wamp/message.py 
    * ``[ERROR, REQUEST.Type|int, REQUEST.Request|id, Details|dict, Error|uri, Payload|binary]``
    * ``[PUBLISH, Request|id, Options|dict, Topic|uri, Payload|binary]``
    * ``[EVENT, SUBSCRIBED.Subscription|id, PUBLISHED.Publication|id, Details|dict, PUBLISH.Payload|binary]``
    * ``[CALL, Request|id, Options|dict, Procedure|uri, Payload|binary]``
    * ``[RESULT, CALL.Request|id, Details|dict, Payload|binary]``
    * ``[INVOCATION, Request|id, REGISTERED.Registration|id, Details|dict, Payload|binary]``
    * ``[YIELD, INVOCATION.Request|id, Options|dict, Payload|binary]``
```

ecorm commented:
While I was consulting the thesaurus, this scene from Lord of the Rings kept playing through my head: "Get back scum! The prisoners go back to Saruman alive and unspoiled."

oberstet commented:
Two more notes rgd feature announcement and consistent use thereafter:

What needs to be announced is probably not only *Protected Payload Mode*, but the specific variant, and in determine all of:
* crypto algo used, eg xsalsa20poly1305
* key management scheme used, eg eth
* serializer used (before encryption), eg cbor

And when a specific flavor is to be used, such as `eth.xsalsa20poly1305.cbor`, then **all** WAMP messages in that session must use the same?

If we want to relax that, then we need to state how this works .. eg at least, `RESULT.details.enc_algo == CALL.details.enc_algo` and similar must hold?

> fwiw, in AB/CB all demos/examples use one schema for everything ...

oberstet commented:
> Get back scum! The prisoners go back to Saruman alive and unspoiled.

=) sounds interesting! unfortunately, haven't seen lord of the rings, and internet search doesn't turn up the respective sequence. I will die ignorant;) or unspoiled? haha

oberstet commented:
to make the decision making / voting easier, let me try to summarize the two options we have rgd the attributes (as I understand what's been suggested that is):

A)

* `enc_serializer`, e.g. `"cbor"`: payload (before enc) serializer
* `enc_algo`, e.g. `"xsalsa20poly1305"`: payload cipher
* `enc_kms`, e.g `"eth"`: key management schema (which also determines the key ID string format)
* `enc_key`, e.g. `"0xe59C7418403CF1D973485B36660728a5f4A8fF9c"`: key ID according to format and semantics determined by `enc_kms`

OR

B)
* `enc_algo`: all 3 above combined, eg `eth.xsalsa20poly1305.cbor`
* `enc_key`: same as above

> fwiw, I'd be fine with both, and I would also be fine with renaming `enc_key` (as it is not the key itself, but an ID pointing to a key) ... it's just that it'll result in tedious work on AB/CB side;)

ecorm commented:
> =) sounds interesting! unfortunately, haven't seen lord of the rings, and internet search doesn't turn up the respective sequence. I will die ignorant;) or unspoiled? haha

https://www.youtube.com/watch?v=ufFOghMt1yI

oberstet commented:
> https://www.youtube.com/watch?v=ufFOghMt1yI

thanks!! so "unspoiled" is about the "meal" consisting of those poor guys. I see. interesting association while looking in the thesaurus rgd "opaque"=) fwiw, when I think of "opaque" and cinema, first thing that comes to my mind is andrei tarkowsky, eg stalker ... "the zone" is definitely opaque https://www.youtube.com/watch?v=YuOnfQd-aTw

om26er commented:
I guess we should update the title of this issue to be `WIP: Protected Payload Mode` ;-)

KSDaemon commented:
Ooookay! Next turn! 

* Renamed feature
* Return enc_key to original meaning
* Added enc_kms options
* Updated some descriptions
* Added info about supporting cyphers and serializers

2 all: Please have a look!

KSDaemon commented:
> to make the decision making / voting easier, let me try to summarize the two options we have rgd the attributes (as I understand what's been suggested that is):
> 
> A)
> B)
> C) as today in AB/CB

Well, I prefer A variant. It's more clear for end user to determine everything. In case of B) — we move some logic to users shoulders. I don't think that they want it :)

So I described A) in spec :)

Jopie64 commented:
> haha! Feeling so stupid))))) Really))) Transparency.... And we are discussing encryption%)))

I thought it was called 'transparent' as in, the router passes the payload transparently. Like, it doesn't even see it, like a window...

oberstet commented:
> > haha! Feeling so stupid))))) Really))) Transparency.... And we are discussing encryption%)))
> 
> I thought it was called 'transparent' as in, the router passes the payload transparently. Like, it doesn't even see it, like a window...

In fact, both is true:

1. from a client point of view, the router treats the payload in a transparent manner .. as if it wasn't there at all (well, mostly, apart from size limits or such)
2. from a router point of view, the payload is treated opaquely .. it cannot look inside .. it (the payload from a router PoV) is opaque

So it is just a dual perspective.

Funny: I was thinking more like you .. from an app PoV ... but I'm fine with both. We just need to pick one.

KSDaemon commented:
> I thought it was called 'transparent' as in, the router passes the payload transparently. Like, it doesn't even see it, like a window...

Well, if in long it reads as `payload transparency to the router` — then why not!?)

But anyway, i think current chosen term is better!

oberstet commented on rfc/text/advanced/ap_50_protected_payload_mode.md:
```diff
@@ -0,0 +1,284 @@
+
+This needs can be covered with `Protected Payload Mode` feature. This feature allows:
+
+* Specifying additional attributes during `CALL`, `PUBLISH`, `EVENT`, `YIELD`, `RESULT` messages 
```
we also need ERROR and INVOCATION

oberstet commented on rfc/text/advanced/ap_50_protected_payload_mode.md:
```diff
@@ -0,0 +1,284 @@
+and *Brokers* (`role := "broker"`) via
+
+{align="left"}
+HELLO.Details.roles.<role>.features.protected_payload_mode|bool := true
```
so this doesn't announce the specific way it will be used in the client, e.g. it doesn't say

```
protected_payload_mode = "eth.xsalsa20poly1305.cbor"
```

or

```
protected_payload_mode = "mqtt.native"
```
or

```
protected_payload_mode = ["eth.xsalsa20poly1305.cbor", "eth.xsalsa20poly1305.flatbuffers"]
```

ecorm commented:
> I thought it was called 'transparent' as in, the router passes the payload transparently. Like, it doesn't even see it, like a window...

> In fact, both is true:
>
>    from a client point of view, the router treats the payload in a transparent manner .. as if it wasn't there at all (well, mostly, apart from size limits or such)
>    from a router point of view, the payload is treated opaquely .. it cannot look inside .. it (the payload from a router PoV) is opaque

"Transparent" is often misused to mean "invisible", as in a process that's hidden and done automatically. But "transparent" really means being able to see through it (like a jar made out of glass). The glass jar is not invisible, and you can see the details of what's inside it.

Consider "opaque types" in C/C++, in which the data structure details are completely hidden (you don't even know its size) and is passed around with a `void` pointer. Considering that "opaque" means the opposite of "transparent", then "transparent" doesn't make sense to describe a mechanism where details are ignored or hidden.

Check out this SO answer, with the highly upvoted comment that reflects my own opinion: https://stackoverflow.com/a/17385138/245265

ecorm commented:
While we're still bikeshedding the name of the feature, I just though of another alternative: "Payload Passthru" or "Payload Passthru Mode".

"Passthru" is used in digital audio to mean that the signal is passed through unchanged to the output.

meejah commented:
While scanning here (who doesn't love a nice "naming" bikeshed!) I was also going to suggest "payload pass-through" (maybe the "passthru" spelling is fine too) because:

- it doesn't _require_ encryption
- says precisely what it means (the server "passes through" the payload, without touching it)
- cases that _do_ use encryption can say "encrypted payload pass-through" for maximum clarity

oberstet commented:
"Payload Passthru": yeah, this is even better .. I also like the "Encrypted Payload Passthru" to specify the variants for that specific use case of it. @KSDaemon what's your take? should you agree, this should be the last change, promise;)

KSDaemon commented:
@oberstet I agree and just renamed it already :)

KSDaemon commented:
2 All: I renamed, removed, added, and edited different sections. Please have a look!

oberstet commented:
@KSDaemon this already looks quite good IMO! also, I really like the new name "Payload-Passthru"=) I still wanted to have a detailed read of the text in the PR, but I think we're "almost there". fwiw, I had a deep look into the respective code in Autobahn and Crossbar.io building on this feature, and I am certain we could match this spec quite easily.

Speaking of _using_ the "Payload-Passthru", e.g. for MQTT and for E2E (v2), I came up with the following precise attribute names and values mapping which we would do:

[wamp-ap-payload-passthru-schemes.pdf](https://github.com/wamp-proto/wamp-proto/files/9222459/wamp-ap-payload-passthru-schemes.pdf)

The idea is: the "Payload-Passthru" just defines the general working, plus provides a mapping table for predefined specific schemes. But those are then to be described in other sections.

Custom schemes can then reuse the basic machinery, as in: a router/client library might want to provide hooks for users to plugin their custom schemes.

Also, as you will note, the attributes are prefixed `ppt_` (for Payload-PassThru), because with MQTT, there isn't any encryption, but there is a special serializer "mqtt" which signals that the payload is the raw MQTT payload (only the payload, not the whole message). So I think this ppt prefix makes more sense.

What do you think?

KSDaemon commented:
@oberstet Looks really good! I agree with renaming. Now it reflects any possible use cases! I'll update the spec accordingly

oberstet commented:
one typo: the following messages have `Details|dict` not `Options|dict`:

* RESULT
* INVOCATION
* ERROR
* EVENT

KSDaemon commented:
> one typo: the following messages have `Details|dict` not `Options|dict`:
> 
> * RESULT
> * INVOCATION
> * ERROR
> * EVENT

My bad. Thnx! Fixed!

oberstet commented:
@KSDaemon awesome, this is perfect! for me, this is ready to merge .. @ecorm what do you think?

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+**ppt_serializer attribute**
+
+`ppt_serializer` attribute is optional. It specifies what serializer was used to build up payload object.
+It can be `mqtt`, `amqp`, `stomp` value just to inform that coming data is from related technologies or ordinary 
```
I don't think it's a good idea for the same attribute to be "overloaded" with two different meanings. Perhaps there should be a separate attribute for things such as `mqtt`, `amqp`, `stomp`, etc.

KSDaemon replied:
Well, maybe... But i think it is a rare case anyway... 
We can think and discuss it...

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+**ppt_scheme attribute**
+
+`ppt_scheme` stands for Key Management Schema. It is required string attribute. This attribute can contain name or
```
What is payload passthru mode is desired in a situation that doesn't involve encryption?

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+**ppt_cipher attribute**
+
+`ppt_cipher` attribute is optional. It is required if payload is encrypted. This attribute specifies cryptographic
```
```suggestion
The `ppt_cipher` attribute is optional. It is required if the payload is encrypted. This attribute specifies the cryptographic
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+**ppt_cipher attribute**
+
+`ppt_cipher` attribute is optional. It is required if payload is encrypted. This attribute specifies cryptographic
+algorithm that was used to encrypt payload. It can be `xsalsa20poly1305`, `aes256gcm` for now.
```
```suggestion
algorithm that was used to encrypt the payload. It can be `xsalsa20poly1305`, `aes256gcm` for now.
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+**ppt_keyid attribute**
+
+`ppt_keyid` attribute is optional. This attribute can contain encryption key id that was used to encrypt payload.
```
```suggestion
The `ppt_keyid` attribute is optional. This attribute can contain the encryption key id that was used to encrypt the payload.
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+**ppt_keyid attribute**
+
+`ppt_keyid` attribute is optional. This attribute can contain encryption key id that was used to encrypt payload.
+`ppt_keyid` attribute is always a type of string. The value can be a hex-encoded string, uri, DNS name, 
```
```suggestion
The `ppt_keyid` attribute is a string type. The value can be a hex-encoded string, URI, DNS name, 
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+`ppt_keyid` attribute is optional. This attribute can contain encryption key id that was used to encrypt payload.
+`ppt_keyid` attribute is always a type of string. The value can be a hex-encoded string, uri, DNS name, 
+Ethereum address, UUID identifier - any meaningful value by which target peer can choose private key 
```
```suggestion
Ethereum address, UUID identifier - any meaningful value which allows the target peer to a choose private key 
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+`ppt_keyid` attribute is optional. This attribute can contain encryption key id that was used to encrypt payload.
+`ppt_keyid` attribute is always a type of string. The value can be a hex-encoded string, uri, DNS name, 
+Ethereum address, UUID identifier - any meaningful value by which target peer can choose private key 
+without guessing. Format of the value may depend on `ppt_scheme` attribute.
```
```suggestion
without guessing. The format of the value may depend on the `ppt_scheme` attribute.
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+Ethereum address, UUID identifier - any meaningful value by which target peer can choose private key 
+without guessing. Format of the value may depend on `ppt_scheme` attribute.
+
+**ppt_ predefined schemes**
```
```suggestion
**ppt_ Predefined Schemes**
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+**ppt_ predefined schemes**
+
+| Attribute      | Required? | Predefined Scheme (mqtt)  | Predefined Scheme (End-to-End Encryption) | Custom Scheme Example |
```
```suggestion
| Attribute      | Required? | Predefined Scheme (MQTT)  | Predefined Scheme (End-to-End Encryption) | Custom Scheme Example |
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+| Attribute      | Required? | Predefined Scheme (mqtt)  | Predefined Scheme (End-to-End Encryption) | Custom Scheme Example |
+|----------------|-----------|---------------------------|-------------------------------------------|-----------------------|
+| ppt_scheme     | Y         | mqtt                      | wamp.eth                                  | x_my_ppt              |
```
```suggestion
| ppt_scheme     | Y         | MQTT                      | wamp.eth                                  | x_my_ppt              |
```

KSDaemon replied:
The same here: we specify values in lowercase

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+| Attribute      | Required? | Predefined Scheme (mqtt)  | Predefined Scheme (End-to-End Encryption) | Custom Scheme Example |
+|----------------|-----------|---------------------------|-------------------------------------------|-----------------------|
+| ppt_scheme     | Y         | mqtt                      | wamp.eth                                  | x_my_ppt              |
+| ppt_serializer | N         | json, msgpack, cbor, mqtt | cbor, flatbuffers                         | custom                |
```
```suggestion
| ppt_serializer | N         | JSON, Msgpack, CBOR, MQTT | cbor, flatbuffers                         | custom                |
```

KSDaemon replied:
I think about skipping this, as it shows a value that we use. And under a hidden agreement the values are lowercase

ecorm replied:
Sorry, I thought you were naming those technologies. I see now that they are identifiers, so yes lowercase.

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+| ppt_cipher     | N         | -                         | xsalsa20poly1305, aes256gcm               | custom                |
+| ppt_keyid      | N         | -                         | *                                         | custom                |
+
+*: Least significant 20 bytes (160 bits) of SHA256 of public key (32 bytes) of data encryption key as HEX encoded 
```
```suggestion
*: The least significant 20 bytes (160 bits) of the SHA256 of the public key (32 bytes) of the data encryption key, as a hex-encoded 
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+| ppt_keyid      | N         | -                         | *                                         | custom                |
+
+*: Least significant 20 bytes (160 bits) of SHA256 of public key (32 bytes) of data encryption key as HEX encoded 
+string with prefix `0x`, and upper/lower case letters encoding a checksum according to EIP55.
```
```suggestion
string with prefix `0x` and either uppercase/lowercase alphabetic characters, encoding a checksum according to EIP55.
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+*: Least significant 20 bytes (160 bits) of SHA256 of public key (32 bytes) of data encryption key as HEX encoded 
+string with prefix `0x`, and upper/lower case letters encoding a checksum according to EIP55.
+
+**Message structure**
```
```suggestion
**Message Structure**
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+**Message structure**
+
+With `Payload Passthru Mode` in use message payload MUST BE sent as one binary item inside 
```
```suggestion
When `Payload Passthru Mode` is in use, the message payload MUST be sent as one binary item within 
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+**Message structure**
+
+With `Payload Passthru Mode` in use message payload MUST BE sent as one binary item inside 
+`Arguments|list`, while `ArgumentsKw|dict` MUST BE missing or empty.
```
```suggestion
`Arguments|list`, while `ArgumentsKw|dict` MUST be absent or empty.
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+*Example.* Caller-to-Dealer progressive `CALL` with encryption and key ID.
+
+Nothing prevents to use `Payload Passthru Mode` with other features, for example `Progressive Calls`
```
```suggestion
Note that nothing prevents the use of `Payload Passthru Mode` with other features such as, for example, `Progressive Calls`
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
````diff
@@ -0,0 +1,401 @@
+    ]
+```
+
+*Example.* Caller-to-Dealer `CALL` with mqtt payload. Specifying `"ppt_serializer": "mqtt"` means, that 
````
```suggestion
*Example.* Caller-to-Dealer `CALL` with MQTT payload. Specifying `"ppt_serializer": "mqtt"` means that 
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
````diff
@@ -0,0 +1,401 @@
+```
+
+*Example.* Caller-to-Dealer `CALL` with mqtt payload. Specifying `"ppt_serializer": "mqtt"` means, that 
+original mqtt message payload is passed as WAMP payload message without any transcoding, as is.
````
```suggestion
the original mqtt message payload is passed as a WAMP payload message as is, without any transcoding.
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
````diff
@@ -0,0 +1,401 @@
+    ]
+```
+
+*Example.* Caller-to-Dealer `CALL` with mqtt payload. Specifying `"ppt_scheme": "mqtt"` simply indicates, that
````
```suggestion
*Example.* Caller-to-Dealer `CALL` with MQTT payload. Specifying `"ppt_scheme": "mqtt"` simply indicates that
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
````diff
@@ -0,0 +1,401 @@
+```
+
+*Example.* Caller-to-Dealer `CALL` with mqtt payload. Specifying `"ppt_scheme": "mqtt"` simply indicates, that
+original source of payload data is received from related system. Specifying `"ppt_serializer": "json"` means, that
````
```suggestion
the original source of payload data is received from a related system. Specifying `"ppt_serializer": "json"` means that
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+*Example.* Caller-to-Dealer `CALL` with mqtt payload. Specifying `"ppt_scheme": "mqtt"` simply indicates, that
+original source of payload data is received from related system. Specifying `"ppt_serializer": "json"` means, that
+original mqtt message payload was parsed and encoded with `json` serializer.
```
```suggestion
the original MQTT message payload was parsed and encoded with the `json` serializer.
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
````diff
@@ -0,0 +1,401 @@
+    ]
+```
+
+*Example.* Dealer-to-Callee `INVOCATION` with mqtt payload
````
```suggestion
*Example.* Dealer-to-Callee `INVOCATION` with MQTT payload
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+*Example.* Callee-to-Dealer progressive `YIELD` with encryption and key ID
+
+Nothing prevents to use `Payload Passthru Mode` with other features, for example `Progressive Call Results`
```
```suggestion
Nothing prevents the use of `Payload Passthru Mode` with other features such as, for example, `Progressive Call Results`
```

ecorm replied:
This is sentence is repeated multiple times through the examples. It should be consolidated in one place.

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+*Example.* Dealer-to-Caller progressive `RESULT` with encryption and key ID
+
+Nothing prevents to use `Payload Passthru Mode` with other features, for example `Progressive Call Results`
```
```suggestion
Nothing prevents the use of `Payload Passthru Mode` with other features such as, for example, `Progressive Call Results`.
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
````diff
@@ -0,0 +1,401 @@
+    ]
+```
+
+**About supported serializers and cryptographic ciphers**
````
```suggestion
**About Supported Serializers and Cryptographic Ciphers**
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+**About supported serializers and cryptographic ciphers**
+
+WAMP works as infrastructure for delivering messages between peers. Regardless of what encryption algorithm 
```
```suggestion
WAMP serves as infrastructure for delivering messages between peers. Regardless of what encryption algorithm 
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+**About supported serializers and cryptographic ciphers**
+
+WAMP works as infrastructure for delivering messages between peers. Regardless of what encryption algorithm 
+and serializer were chosen for `Payload Passthru Mode`, *Router* will not inspect and analyze related encryption 
```
```suggestion
and serializer were chosen for `Payload Passthru Mode`, a *Router* shall not inspect and analyze the `ppt_` options and payload
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+
+WAMP works as infrastructure for delivering messages between peers. Regardless of what encryption algorithm 
+and serializer were chosen for `Payload Passthru Mode`, *Router* will not inspect and analyze related encryption 
+message options and payload. It is up to an application side code responsibility to use serializers and 
```
```suggestion
of encypted messages. The application is responsible for choosing serializers and 
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,401 @@
+WAMP works as infrastructure for delivering messages between peers. Regardless of what encryption algorithm 
+and serializer were chosen for `Payload Passthru Mode`, *Router* will not inspect and analyze related encryption 
+message options and payload. It is up to an application side code responsibility to use serializers and 
+ciphers known to every peer involved into message processing.
```
```suggestion
ciphers known to every peer involved in message processing.
```

ecorm reviewed (changes requested):
Grammar fixes and readability suggestions.

ecorm commented:
> @ecorm what do you think

Made a pass to fix grammar and improve readability, with about a million missing "a" or "the" articles :wink: (yes I understand they're not used in Slavic languages, just poking fun).

I have two technical critiques:
- The `ppt_scheme` attribute is required to be non-empty, and its presence is what "triggers" Payload Passthru Mode. What if an application wants to enable Payload Passthru Mode, but doesn't want/need to use encryption?
- I don't like that `ppt_serializer` is "overloaded" for two different meanings: identifying the message transport technology (MQTT, etc) or identifying the actual serializer. Could it be split into two attributes?

ecorm commented:
Another technical critique is that there needs to be more clarification about what to do when the WAMP-level messages are serialized as JSON. I'm guessing the payload should be Base64-encoded, but this needs to be stated somewhere explicitly (unless I missed it).

Another corner case that needs clarification is when WAMP-level messages are JSON, and passthru payloads are also JSON.

KSDaemon commented:
Phew!! I've done it!)))

KSDaemon commented:
@ecorm 
> The ppt_scheme attribute is required to be non-empty, and its presence is what "triggers" Payload Passthru Mode. What if an application wants to enable Payload Passthru Mode, but doesn't want/need to use encryption?
In that case app should just specify some value there. Maybe we should explicitly add `passthru` value for such cases? 
@oberstet What do you think?

> I don't like that ppt_serializer is "overloaded" for two different meanings: identifying the message tunneling technology (MQTT, etc) or identifying the actual serializer. Could it be split into two attributes?

For identifying tech there is a `ppt_scheme` attribute. So serializer attribute is only for specifying serializers in my opinion.

KSDaemon commented:
> Another technical critique is that there needs to be more clarification about what to do when the WAMP-level messages are serialized as JSON. I'm guessing the payload should be Base64-encoded (with nul character prefix), but this needs to be stated somewhere explicitly (unless I missed it).
> 
> Another corner case that needs clarification is when WAMP-level messages are JSON, and passthru payloads are also JSON.

These are  great questions!

KSDaemon commented:
Tried to find some words in spec about encoding binary data into JSON... No luck...
So.. Yeah, I agree, let's explicitly say about base64 encoding of binary data when JSON serializer is used in WAMP session.

Regarding JSON-in-JSON — we can state that payload may be string or binary or we can specify the requirement for prepending null character for example to JSON-string...

oberstet commented:
>     * The `ppt_scheme` attribute is required to be non-empty, and its presence is what "triggers" Payload Passthru Mode. What if an application wants to enable Payload Passthru Mode, but doesn't want/need to use encryption?

then the application can use `ppt_scheme` without using `ppt_cipher` and `ppt_keyid`

>     * I don't like that `ppt_serializer` is "overloaded" for two different meanings: identifying the message tunneling technology (MQTT, etc) or identifying the actual serializer. Could it be split into two attributes?

ok, I can follow that argument. however, in this case, how about the following? which doesn't intro a new attribute:

```
ppt_scheme == "mqtt", ppt_serializer == "native"
```

for specifying MQTT, with a "native" payload, that is native and custom to the application ("native" in as: the WAMP client library will also not touch it .. deliver raw bytes to/from the app code, rather than applying the specified `ppt_serializer` before delivering to/from app code)

-----

> Another technical critique is that there needs to be more clarification about what to do when the WAMP-level messages are serialized as JSON.

so when `ppt_serializer` is specified `!= "mqtt/native"`, eg `"json"`, then the serializer is applied first (by the WAMP client library), and then (without encryption), the resulting binary payload is embedded in a WAMP message, which might use a different serializer.

if that 2nd WAMP message level serializer is JSON, then the implementation must support

https://github.com/wamp-proto/wamp-proto/blob/master/rfc/text/advanced/ap_41_serialization_binary_json.md

however, for some reasons, it seems this file isn't currently referenced in the spec top-level texts, and hence not rendered. which is a bug;)

oberstet commented:
> Yeah, I agree, let's explicitly say about base64 encoding of binary data when JSON serializer is used in WAMP session.

the "Binary support in JSON" section has been currently lost in the spec (pls see above). this is a bug.

however, that still doesn't answer the question of how the binary `payload` resulting from applying `ppt_serializer` is to be treated in `args[0]` in the embedded WAMP message - treated wrt the serializer used for the WAMP message

I would argue: the embedded WAMP message serializer MUST support binary payloads, and for JSON, that means, the implementation MUST support the "Binary support in JSON" feature

I would not want to pay for Base64 encoding of _all_ my WAMP-JSON traffic by _deactivating_ this feature. I mean, yeah, I would use CBOR anyways, but ... - at least this is my argument.

The problem arises necessarily since: the type of `payload` is always binary, but the outer embedded WAMP message serializer might not be able to transport that unencoded

KSDaemon commented:
@oberstet @ecorm I added more text covering what we discussed above. Yep, that is not completed, but please have a look and maybe provide some feedback or additions to this. Let's get together and make one last push!

KSDaemon commented:
@ecorm Your thoughts on the last updates?

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+
+**ppt_ Predefined Schemes**
+
+| Attribute      | Required? | Predefined Scheme (MQTT)    | Predefined Scheme (End-to-End Encryption) | Custom Scheme Example |
```
It's not clear if MQTT and EtoE are just examples or are mandated by the WAMP spec.

ecorm replied:
This would be best split into separate tables to fit the page width in RFC text format.

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+as a hex-encoded string with prefix `0x` and either uppercase/lowercase alphabetic characters, encoding a 
+checksum according to EIP55.
+
+When `Payload Passthru Mode` is used for gateways to other technologies like MQTT Brokers or AMQP Queues and
```
```suggestion
When `Payload Passthru Mode` is used for gateways to other technologies, such as MQTT Brokers or AMQP Queues, then
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+checksum according to EIP55.
+
+When `Payload Passthru Mode` is used for gateways to other technologies like MQTT Brokers or AMQP Queues and
+the `ppt_serializer` attribute may be set to `native` value. This means that payload is not touched by WAMP peers, 
```
```suggestion
the `ppt_serializer` attribute may be set to the `native` value. This means that the payload is not to be modified by WAMP peers, 
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+
+When `Payload Passthru Mode` is used for gateways to other technologies like MQTT Brokers or AMQP Queues and
+the `ppt_serializer` attribute may be set to `native` value. This means that payload is not touched by WAMP peers, 
+not serialized in any manner and is delivered as is from the originator peer. Another possible case is when
```
```suggestion
nor serialized in any manner, and is delivered as-is from the originating peer. Another possible case is when
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+When `Payload Passthru Mode` is used for gateways to other technologies like MQTT Brokers or AMQP Queues and
+the `ppt_serializer` attribute may be set to `native` value. This means that payload is not touched by WAMP peers, 
+not serialized in any manner and is delivered as is from the originator peer. Another possible case is when
+the `ppt_serializer` attribute is set to any valid serializer, for example `msgpack`. In this case WAMP originator
```
```suggestion
the `ppt_serializer` attribute is set to any valid serializer, for example `msgpack`. In this case, the originating WAMP
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+the `ppt_serializer` attribute may be set to `native` value. This means that payload is not touched by WAMP peers, 
+not serialized in any manner and is delivered as is from the originator peer. Another possible case is when
+the `ppt_serializer` attribute is set to any valid serializer, for example `msgpack`. In this case WAMP originator
+client peer first applies `ppt_serializer` to serialize payload (without encryption), then the resulting
```
```suggestion
client peer first applies `ppt_serializer` to serialize the payload (without encryption), then the resulting
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+not serialized in any manner and is delivered as is from the originator peer. Another possible case is when
+the `ppt_serializer` attribute is set to any valid serializer, for example `msgpack`. In this case WAMP originator
+client peer first applies `ppt_serializer` to serialize payload (without encryption), then the resulting
+binary payload is embedded in a WAMP message, which might use different serializer that was chosen during
```
```suggestion
binary payload is embedded in the WAMP message, the latter having possibly a different serializer depending on the one chosen during
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+
+**Important note regarding JSON serializer**
+
+With `Payload Passthru Mode` payloads are treated as binaries. To send binaries WAMP session serializer MUST
```
```suggestion
With `Payload Passthru Mode`, payloads are treated as binary. To send these binary payloads, the WAMP session serializer MUST
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+**Important note regarding JSON serializer**
+
+With `Payload Passthru Mode` payloads are treated as binaries. To send binaries WAMP session serializer MUST
+support binary types. Most of the serializers support binary types, but JSON serializer does not. To use
```
```suggestion
support byte arrays. Most serialization formats known to WAMP support byte arrays, but JSON does not support them natively. To use
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+
+With `Payload Passthru Mode` payloads are treated as binaries. To send binaries WAMP session serializer MUST
+support binary types. Most of the serializers support binary types, but JSON serializer does not. To use
+`Payload Passthru Mode` with JSON serializer WAMP peers MUST support 
```
```suggestion
`Payload Passthru Mode` with a JSON serializer, WAMP peers MUST perform the special
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+With `Payload Passthru Mode` payloads are treated as binaries. To send binaries WAMP session serializer MUST
+support binary types. Most of the serializers support binary types, but JSON serializer does not. To use
+`Payload Passthru Mode` with JSON serializer WAMP peers MUST support 
+[Binary serialization in JSON](#binary-support-in-json). This come with huge overhead, so general advice is
```
```suggestion
[Binary serialization in JSON](#binary-support-in-json). This conversion may have unacceptable overhead, so it is generally advised
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+support binary types. Most of the serializers support binary types, but JSON serializer does not. To use
+`Payload Passthru Mode` with JSON serializer WAMP peers MUST support 
+[Binary serialization in JSON](#binary-support-in-json). This come with huge overhead, so general advice is
+to use WAMP session serializers with native binary support, for example `MessagePack`, `CBOR`, `FlatBuffers`.
```
```suggestion
to use WAMP session serializers with native byte array support, for example, `MessagePack`, `CBOR`, or `FlatBuffers`.
```

ecorm commented on rfc/text/advanced/ap_50_payload_passthru_mode.md:
```diff
@@ -0,0 +1,422 @@
+binary payload is embedded in a WAMP message, which might use different serializer that was chosen during
+WAMP Session establishment.
+
+**Important note regarding JSON serializer**
```
```suggestion
**Important Note Regarding JSON Serialization**
```

ecorm reviewed (changes requested):
Grammar and readability of latest changes concerning serializers.

KSDaemon commented:
Okay! Fixed that! Hope it was the last round %)

KSDaemon commented:
OMG! It is merged! 🙌🏻
